docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,364 @@
+     This timeout is typically much shorter than the deduplication duration.
+
+   .. _dedup-bounded-step-await:
+   
+#. Wait until the RPC call returns a response.
+   
+   - Status codes other than ``OK`` should be handled according to :ref:`error handling <dedup-bounded-error-handling>`.
+
+   - When you use the :ref:`Command Service <command-service>` and the response carries the status code ``OK``, the ledger change took place.
+     You can report success.
+     
+   - When you use the :ref:`Command Submission Service <command-submission-service>`,
+     subscribe with the :ref:`Command Completion Service <command-submission-service>` for completions for ``actAs`` from ``OFF1`` (exclusive) until you see a completion event for the change ID and the submission ID chosen in :ref:`Step 3 <dedup-bounded-step-submit>`.
+     If the completion’s status is ``OK``, the ledger change took place and you can report success.
+     Other status codes should be handled according to :ref:`error handling <dedup-bounded-error-handling>`.
+   
+     This step needs no timeout as the :ref:`Command Submission Service <command-submission-service>` acknowledges a submission only if there will eventually be a completion event, unless relevant parts of the system become permanently unavailable.
+
+
+.. _dedup-bounded-error-handling:
+
+Error handling
+--------------
+
+Error handling is needed when the status code of the command submission RPC call or in the :ref:`in the completion event <com.daml.ledger.api.v1Completion.status>` is not ``OK``.
+The following table lists appropriate reactions by status code (written as ``STATUS_CODE``) and error code (written in capital letters with a link to the error code documentation).
+Fields in the error metadata are written as ``field`` in lowercase letters.
+
+.. list-table:: Command deduplication error handling with known processing time bound
+   :widths: 10 50
+   :header-rows: 1
+
+   - * Error condition
+     
+     * Reaction
+
+       
+   - * ``DEADLINE_EXCEEDED``
+     
+     * Consider the submission lost.
+       
+       Retry from :ref:`Step 2 <dedup-bounded-step-offset>`, "obtaining the completion offset ``OFF1``", and possibly increase the timeout.
+
+       
+   - * Application crashed
+     
+     * Retry from :ref:`Step 2 <dedup-bounded-step-offset>`, "obtaining the completion offset ``OFF1``".
+
+
+   - * ``ALREADY_EXISTS`` / :ref:`DUPLICATE_COMMAND <error_code_DUPLICATE_COMMAND>`
+     
+     * The change ID has already been accepted by the ledger within the reported deduplication period.
+       The optional field ``completion_offset`` contains the precise offset.
+       The optional field ``existing_submission_id`` contains the submission ID of the successful submission.
+       Report success for the ledger change.
+       
+       If desired, query the ``completion_offset`` via the :ref:`Command Completion Service <command-submission-service>` to find out about the earlier outcome.
```

Comment:
It's currently not possible to query a specific `completion_offset`, and if we use the streaming approach the offset is actually exclusive. How would querying work in this case?

Reply:
Good point. I've removed this now. We can add it back in when/if we support querying by offset.